record-layer.go

```diff
@@ -50,10 +62,28 @@ type cipherState struct {
 	cipher   cipher.AEAD // AEAD cipher
 }
 
-type RecordLayer struct {
+type RecordLayerFactory interface {
+	NewLayer(conn io.ReadWriter, dir Direction) RecordLayer
+}
+
+type RecordLayer interface {
+	Lock()
+	Unlock()
+	SetVersion(v uint16)
+	SetLabel(s string)
+	Rekey(epoch Epoch, factory AeadFactory, keys *KeySet) error
+	ResetClear(seq uint64)
+	DiscardReadKey(epoch Epoch)
+	PeekRecordType(block bool) (RecordType, error)
+	ReadRecord() (*TLSPlaintext, error)
+	WriteRecord(pt *TLSPlaintext) error
+	Epoch() Epoch
+}
+
+type RecordLayerImpl struct {
```

Comment:
I would prefer something more descriptive, e.g., `DefaultRecordLayer`.  Since the whole premise here is that there can be other impls.

Reply:
We should also actually make this the default impl, so that if a record layer isn't specified, this one gets filled in.  Unclear to me from this PR whether this is already the case.

Reply:
Yeah, this looks like it is wrong to me.  The number of casts down to `DefaultRecordLayer` are too high.  If the intent is to abstract this out, then this set of changes falls short.  I've manage to fix the other comments, but this one is tricky.